File: leetcode/ContainsDuplicate/containsDuplicate.py
# Solution: Using a HashTable to store a frequency table.
# Complexity: Time: O(n) Space: O(n)
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        countDict = {}
        for i in range(len(nums)):
            if nums[i] in countDict:
                countDict[nums[i]] += 1
            else:
                countDict[nums[i]] = 1
        for key in countDict:
            if countDict[key] > 1:
                return True

# Runtime: 136 ms, faster than 25.43% of Python3 online submissions for Contains Duplicate.
# Memory Usage: 20 MB, less than 7.55% of Python3 online submissions for Contains Duplicate.

# Solution: Using a Hash Table an returning true if the value is already in there
# Complexity: Time: O(n) Space: O(n)
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        countDict = {}
        for i in range(len(nums)):
            if nums[i] in countDict:
                return True
            else:
                countDict[nums[i]] = 1
        return False
# Runtime: 132 ms, faster than 44.17% of Python3 online submissions for Contains Duplicate.
# Memory Usage: 19.2 MB, less than 16.98% of Python3 online submissions for Contains Duplicate.


# Solution: Using set 

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        return (len(nums) > len(set(nums)))

# Runtime: 116 ms, faster than 97.91% of Python3 online submissions for Contains Duplicate.
# Memory Usage: 18.3 MB, less than 88.68% of Python3 online submissions for Contains Duplicate.

File: leetcode/ContainsDuplicate/containsDuplicate.js
// Solution: Build a frequency table
// Complexity: Time O(n) | Space O(n)
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function(nums) {
    let map = new Map();
    const n = nums.length;
    //Needed because in a forEach method, the only way to exit is throwing an exception
    let answer = false;
    for (let i = 0; i < n; i++){
        if(map.has(nums[i])){
            const iFreq = map.get(nums[i]) + 1;
            map.set(nums[i], iFreq);   
        }
        else{
            map.set(nums[i], 1);
        }
    }
    map.forEach((value, key)=>{
        if(value > 1)
            answer = true;
    });
    return answer;
};

// Runtime: 64 ms, faster than 83.25% of JavaScript online submissions for Contains Duplicate.
// Memory Usage: 42.5 MB, less than 17.65% of JavaScript online submissions for Contains Duplicate.

// Solution: Build a Hash Table 
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var containsDuplicate = function(nums) {
    let map = new Map();
    const n = nums.length;
    for (let i = 0; i < n; i++){
        if(map.has(nums[i])){
           return true;   
        }
        else{
            map.set(nums[i], 1);
        }
    }
    return false;
};

//Runtime: 68 ms, faster than 71.41% of JavaScript online submissions for Contains Duplicate.
//Memory Usage: 41.1 MB, less than 61.76% of JavaScript online submissions for Contains Duplicate.
/*
There are many data structures commonly used as dynamic sets such as Binary Search Tree 
and Hash Table. The operations we need to support here are search() and insert(). For a 
self-balancing Binary Search Tree (TreeSet or TreeMap in Java), search() and insert() are 
both O(\log n)O(logn) time. For a Hash Table (HashSet or HashMap in Java), search() and insert() 
are both O(1)O(1) on average. Therefore, by using hash table, we 
can achieve linear time complexity for finding the duplicate in an unsorted array. */
